### PR TITLE
utxonursery: add test coverage

### PR DESCRIPTION
**Pull request by joostjager**
This PR adds test coverage to utxonursery.

In addition to this it fixes a bug where IncubateOutputs doesn't add to the waitgroup and might leave lingering goroutines after nursery is stopped.

### CODE REVIEW

**Comment by joostjager** [comment]
@cfromknecht Yes, db field needed to be removed in order to mock the call.

File: utxonursery_test.go [hunk 1]
```diff
@@ -379,3 +387,707 @@ func TestBabyOutputSerialization(t *testing.T) {
...
+
+func (m *nurseryMockNotifier) Stop() error {
+	return nil
+}
+
+func (m *nurseryMockNotifier) RegisterSpendNtfn(outpoint *wire.OutPoint,
+	_ []byte, heightHint uint32) (*chainntnfs.SpendEvent, error) {
+
+	return &chainntnfs.SpendEvent{
+		Spend:  m.spendChan,
```
[the remark this answers is not in the document]
Comment (joostjager) [reply utxonursery_test.go]: In this PR it is just to satisfy the interface. I still have more test code that also exercises the not yet present remote spend functionality of nursery and that does use this channel. Do you want me to return nil here, or shall we leave it in as is (that is what I prefer)?

File: utxonursery_test.go [hunk 2]
```diff
@@ -379,3 +387,708 @@ func TestBabyOutputSerialization(t *testing.T) {
...
+			currentStartStopIdx++
+			if restartIdx == currentStartStopIdx {
+				startStopFunc()
+
+				return true
+			}
+			return false
+		}
+
+		t.Run(fmt.Sprintf("restart_%v", restartIdx),
```
[the remark this answers is not in the document]
Comment (joostjager) [reply utxonursery_test.go]: The return value indicates whether the test succeeded or not. In this case, I think it should continue with trying the other restart points and therefore the return value is irrelevant.

File: utxonursery_test.go [hunk 3]
```diff
@@ -379,3 +387,708 @@ func TestBabyOutputSerialization(t *testing.T) {
...
+func newNurseryMockNotifier() *nurseryMockNotifier {
+	return &nurseryMockNotifier{
+		confChannel: make(map[chainhash.Hash]chan *chainntnfs.TxConfirmation),
+		epochChan:   make(chan *chainntnfs.BlockEpoch),
+		spendChan:   make(chan *chainntnfs.SpendDetail),
+	}
+}
+
+func (m *nurseryMockNotifier) confirmTx(txid *chainhash.Hash, height uint32) {
+	m.getConfChannel(txid) <- &chainntnfs.TxConfirmation{
```
[the remark this answers is not in the document]
Comment (joostjager) [reply utxonursery_test.go]: Good point. I've been waiting too many times for timeouts during development. Fixed.

File: utxonursery_test.go [hunk 4]
```diff
@@ -379,3 +387,708 @@ func TestBabyOutputSerialization(t *testing.T) {
...
+// will attempt to execute a restart action in a different location. This is to
+// assert that the unit under test is recovering correctly from restarts.
+func testRestartLoop(t *testing.T, test func(*testing.T,
+	func(func()) bool)) {
+
+	// Start with running the test without any restarts (index zero)
+	restartIdx := 0
+
+	for {
+		currentStartStopIdx := 0
```
[the remark this answers is not in the document]
Comment (joostjager) [reply utxonursery_test.go]: It is indeed a little hard to parse, but this was the way to keep the amount of duplicated code in the tests minimal.

What you describe (running the test twice) is not what is happening. 

The test indeed first runs without restart. I updated the code to actually use `no_restart` as subtest name to make that more clear.

After that, the test is ran `n` more times. `n` is the number of times a call to `ctx.Restart()` is made in the test code. Example output:
```
=== RUN   TestNurseryOutgoingHtlcSuccessOnRemote
=== RUN   TestNurseryOutgoingHtlcSuccessOnRemote/no_restart
=== RUN   TestNurseryOutgoingHtlcSuccessOnRemote/restart_1
=== RUN   TestNurseryOutgoingHtlcSuccessOnRemote/restart_2
=== RUN   TestNurseryOutgoingHtlcSuccessOnRemote/restart_3
=== RUN   TestNurseryOutgoingHtlcSuccessOnRemote/restart_4
=== RUN   TestNurseryOutgoingHtlcSuccessOnRemote/restart_5
=== RUN   TestNurseryOutgoingHtlcSuccessOnRemote/restart_6
--- PASS: TestNurseryOutgoingHtlcSuccessOnRemote (1.01s)
```
So 6+1 times in this case.

It will determine on the fly how many calls to `ctx.Restart()` are present in the code. So if you want to add an extra restart point, there is nothing more to do than just add a `ctx.Restart()` call. The actual restart code that is then executed, is specified in `createNurseryTestContext` (stop/start nursery). Running the test `n` times, keeping counters and detecting the final run, is done in `testRestartLoop`.

File: utxonursery_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (cfromknecht) [reply utxonursery_test.go]: For now I think it's fine to leave as is, no need to modify now if it isn't being used

File: utxonursery_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cfromknecht) [reply utxonursery_test.go]: awesome, thank you

File: utxonursery.go [hunk 5]
```diff
@@ -346,6 +352,13 @@ func (u *utxoNursery) IncubateOutputs(chanPoint wire.OutPoint,
 	incomingHtlcs []lnwallet.IncomingHtlcResolution,
 	broadcastHeight uint32) error {
 
+	// Add to wait group because nursery might shut down during execution of
+	// this function. Otherwise it could happen that nursery thinks it is
+	// shut down, but in this function new goroutines were started and stay
+	// around.
+	u.wg.Add(1)
```
Comment (cfromknecht) [inline utxonursery.go]: Another potential avenue for trying to ensure that we don't shut down too early might be to lock around the closing of `quit` in the stop method. inside the critical section below, we can then check the quit chan to see if we should bail early. I think this may be less racey wrt to modifying the waitgroup counter too late

Comment (joostjager) [reply utxonursery.go]: Added commit that checks for quit channel. I think it is just as safe as checking inside the critical section, but needs less code.
